import "radiant.proto";
import "store.proto";

package radiant.tesseract.protocol;

option optimize_for = SPEED;

message Request {
   extensions 100 to max;
   enum Type {
      PostCommandRequest = 2;
      FinishedUpdate     = 3;
   }
   required Type     type = 1;
};

message PostCommandRequest {
   extend Request {
      optional PostCommandRequest extension = 102;
   }
   enum Op {
      CALL = 1;
      INSTALL_TRACE = 2;
      REMOVE_TRACE = 3;
   }
   required int32    call_id = 1;
   required Op       op = 2;
   optional string   object = 3;
   optional string   route = 4;
   optional string   args = 5;
}

message FinishedUpdate {
   extend Request {
      optional FinishedUpdate extension = 103;
   }
   required uint32   sequence_number = 1;
}

message Update {
   extensions 100 to max;
   enum Type {
      BeginUpdate = 1;
      SetServerTick = 2;
      CommandReply = 3;
      UpdateObject = 4;
      RemoveObjects = 5;
      AllocObjects = 6;
      UpdateDebugShapes = 7;
      UpdateObjectsInfo = 8;
      EndUpdate = 9;
      PostCommandReply = 11;
      LoadGame = 12;
      DefineRemoteObject = 13;
   }
   required Type     type = 1;
   optional int32    chunk_size = 2;
   optional int32    chunk_progress = 3;
   optional int32    reply_id = 4;
}

message BeginUpdate {
   extend Update {
      optional BeginUpdate extension = 101;
   }
   required uint32   sequence_number = 1;
}

message SetServerTick {
   extend Update {
      optional SetServerTick extension = 102;
   }
   required uint32   now = 1;
   required uint32   interval = 2;
   required uint32   next_msg_time = 3;
}

message CommandReply {
   extend Update {
      optional CommandReply extension = 103;
   }
   required int32    reply_id = 1;
   required string   msg = 2;
}

message UpdateObject {
   extend Update {
      optional UpdateObject extension = 104;
   }
   required Protocol.Object object = 1;
}

message RemoveObjects {
   extend Update {
      optional RemoveObjects extension = 105;
   }
   repeated uint32 objects = 1;
}

message AllocObjects {
   extend Update {
      optional AllocObjects extension = 106;
   }
   message Entry {
      required int32 object_id = 1;
      required int32 object_type = 2;
   }
   repeated Entry objects = 1;
}

message UpdateDebugShapes {
   extend Update {
      optional UpdateDebugShapes extension = 107;
   }
   required radiant.protocol.shapelist shapelist = 1;
}

message EndUpdate {
   extend Update {
      optional EndUpdate extension = 109;
   }
}

message LoadGame {
   extend Update {
      optional LoadGame extension = 108;
   }
   optional string         game_id = 1;
}

message PostCommandReply {
   extend Update {
      optional PostCommandReply extension = 120;
   }
   enum Type {
      PROGRESS = 1;
      DONE = 2;
      FAIL = 3;
   }
   required int32       call_id = 1;
   optional Type        type = 2;
   optional bytes       content = 3; // compressed json string.
}

message DefineRemoteObject {
   extend Update {
      optional DefineRemoteObject extension = 122;
   }
   optional string      uri = 1;
   optional string      object_uri = 2;
}

message ClientRequest {
   optional int32        request_id = 1;
   required string       request = 2; // json
}

message UpdateObjectsInfo {
   extend Update {
      optional UpdateObjectsInfo extension = 123;
   }
   optional uint32         update_count = 1;
   optional uint32         remove_count = 2;
}

